File: i2c.py
#!/usr/bin/env python3.8
import logging
import json
from time import sleep
from smbus2 import SMBus

logging.basicConfig(format="%(message)s")
log = logging.getLogger(__name__)

# Configure static parameters here for now.
# TODO: implement external config file from json or cmdline params

i2c_bus = 1         # i2c bus to use. default to /dev/i2c-1
sensor_addr = 0x5c  # i2c address of the device
func_code = 0x03    # always 0x03 on AM2320.
reg_len = 0x04      # AM2320 has total 4 registers for humi and temp data.
reg_start = 0x00    # start reading from zero by default


try:
    # could try `with SMBus(1) as bus:`, not sure how it works though
    bus = SMBus(i2c_bus)
except Exception as err:
    log.error(f"Error opening i2c: {err}")


def readSensor(addr=sensor_addr, start=reg_start, op=func_code, reg=reg_len):
    """
    TODO:\n
    Document this!
    """
    bus.write_byte(addr, 0, True)
    # bus.close() might be needed after every operation.
    sleep(0.0850)  # AM2320 needs 500-1000 µs of sleep after wake up

    bus.write_i2c_block_data(addr, start, [op, reg])
    sleep(0.170)  # AM2320 needs to sleep 100-200 ms before reading the registers

    data = bus.read_i2c_block_data(addr, start, reg)
    bus.close()

    # TODO:
    # Implement (or find) crc16 function to verify the data.
    # do not return the value unless crc16 passes.
    # if crc16 does not pass we need to wait xxx ms to read again.

    # C Code below:
    """
    unsigned short crc16(unsigned char *ptr, unsigned char len)
    {
        unsigned short crc = 0xFFFF;
        unsigned char i;
        while (len--)
        {
            crc ^= *ptr++;
            for (i = 0; i < 8; i++)
            {
                if (crc & 0x01)
                {
                    crc >>= 1;
                    crc ^= 0xA001;
                }
                else
                {
                    crc >>= 1;
                }
            }
        }
        return crc;
    }
    """

    # return the data as hex list for now
    return data

File: i2c.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include <time.h>

unsigned short crc16(unsigned char *ptr, unsigned char len)
{
    unsigned short crc = 0xFFFF;
    unsigned char i;
    while (len--)
    {
        crc ^= *ptr++;
        for (i = 0; i < 8; i++)
        {
            if (crc & 0x01)
            {
                crc >>= 1;
                crc ^= 0xA001;
            }
            else
            {
                crc >>= 1;
            }
        }
    }
    return crc;
}

int main()
{
    time_t timer;
    char t_buffer[20];
    struct tm *tm_info;

    int file_i2c;
    int len;
    unsigned char buf[8];

    // open i2c bus in RW mode
    //////////////////////////////////////
    char *fname = (char *)"/dev/i2c-1";

    if ((file_i2c = open(fname, O_RDWR)) < 0)
    {
        printf("failed to open i2c");
        exit(1);
    }

    // open i2c device
    //////////////////////////////////////
    int addr = 0x5c;
    if (ioctl(file_i2c, I2C_SLAVE, addr) < 0)
    {
        printf("failed to communicate with i2c device");
        exit(1);
    }
    usleep(850);

    // wake up sensor by sending something
    //////////////////////////////////////
    len = 1;
    buf[0] = 0x00;
    write(file_i2c, buf, len);

    usleep(850);

    // ask for data
    //////////////////////////////////////
    buf[0] = 0x03; // opening command
    buf[1] = 0x00; // start register
    buf[2] = 0x04; // num registers
    len = 3;
    if (write(file_i2c, buf, len) != len)
    {
        printf("failed to write");
        exit(1);
    }

    usleep(1700);

    // read data and write to file
    //////////////////////////////////////
    FILE *f = fopen("data.txt", "w");
    fflush(f);
    fclose(f);

    FILE *file = fopen("data.txt", "a");
    if (file == NULL)
    {
        printf("Error opening file\n");
        exit(1);
    }

    len = 8;
    if (read(file_i2c, buf, len) != len)
    {
        printf("failed to read i2c bus");
        exit(1);
    }
    else
    {
        int i, temp;
        unsigned short humi, checksum, crc;

        humi = buf[2] << 8;
        humi = humi + buf[3];

        temp = buf[4] << 8;
        temp = temp + buf[5];

        // uncomment to test negative temperature value.
        /*
        temp = temp + 0x8000;
        printf("%04x\n\n", temp);
        */

        // negative temp is determined if bit 15 is 1
        if (temp & 0x8000)
        {
            temp = -(temp & 0x7fff);
        }

        crc = buf[7] << 8;
        crc = crc + buf[6];

        checksum = crc16(buf, 6);

        if (checksum == crc)
        {
            timer = time(NULL);
            tm_info = localtime(&timer);

            strftime(t_buffer, 20, "%Y-%m-%d %H:%M:%S", tm_info);

            printf("RH: %i.2\nTemp: %i.2\nCRC: %02x\n", humi / 10, temp / 10, crc);
            printf("Timestamp: %s\n\n", t_buffer);

            fprintf(file, "\"%i.2\",\"%i.2\",\"%s\"", humi / 10, temp / 10, t_buffer);

            /*
            for (i = 0; i < len; i++)
            {
                printf("%02x ", buf[i]);
            }
*/
            printf("\n");
        }
        else
        {
            printf("Checksum error.\nsensor: %04x\ncrc 16: %04x\n\ndata: ", crc, checksum);

            for (i = 0; i < len; i++)
            {
                printf("%02x ", buf[i]);
            }
            printf("\n");

            exit(1);
        }
    }

    fclose(file);
    sleep(0.05);

    return 0;
}
